### ISSUE TITLE
Confusing BCF error messages, and mixing warnings vs errors.

### ISSUE THREAD

**Issue by jkbonfield** [samtools/htslib#1359]
If I have a VCF file with a missing `##FORMAT` record, then a straight `bcftools view` issues a warning:

```
@ seq4-head2[tmp/vcf]; bcftools view passed_ploidy_001.vcf|wc
[W::vcf_parse_format] FORMAT 'PL' at 1:61462 is not defined in the header, assuming Type=String
     14      81     759
```

If I try to output the same file as BCF however, I get the warning and then a boggling error code:

```
@ seq4-head2[tmp/vcf]; bcftools view -Ob passed_ploidy_001.vcf |wc
[W::vcf_parse_format] FORMAT 'PL' at 1:61462 is not defined in the header, assuming Type=String
[E::bcf_write] Unchecked error (2) at 1:61462
[main_vcfview] Error: cannot write to (null)
      0       0       0
```

What does "Unchecked error (2)" mean?  Initially I didn't associate this with the previous warning because we get those on VCF too and all is fine, but it's actually that which is now fatal.  If it's actually an error to lack the FORMAT header then it should be reported as an error when outputting BCF and not a warning.

**Comment by jkbonfield** [comment]
Of course I can answer all these questions by looking at the source.  Error 2 comes from these in htslib/vcf.h

```
#define BCF_ERR_CTG_UNDEF 1
#define BCF_ERR_TAG_UNDEF 2
#define BCF_ERR_NCOLS     4
#define BCF_ERR_LIMITS    8
#define BCF_ERR_CHAR     16
#define BCF_ERR_CTG_INVALID   32
#define BCF_ERR_TAG_INVALID   64
```

I suggest we add a `vcf_strerror` function to convert the bit-field into a series of error strings, and use that in printing up the messages.  It's pretty unfair on uses to simply quote a numeric at them.  It's also not clear what "Unchecked" means in this context.  Again I can get it by wading through code, but we need a better term.

**Comment by pd3** [comment]
The tag PL was encountered in the VCF body but was not defined in the header. When such tag is encountered, the reader can decide to abort (but that would render many VCFs unreadable) or to go with a made up type of annotation (the most flexible is string and can cover anything). When writing such tag out, with VCF output it is not a problem, it's just written out as a string. With BCF, however, this leads to a critical error, the BCF dicitonary (i.e. the VCF header) has been already printed and cannot be modified.

As for the error reporting, `vcf_strerror` is a good idea indeed.

**Comment by jkbonfield** [comment]
I understand the reasoning, but it still needs to be an "E:" and not a "W:" output as this is now fatal.

**Comment by pd3** [comment]
No. The first is printed while reading and can be recovered from, hence a warning. It does not know that the record will be later used for writing. Only then it becomes an error.

**Comment by jkbonfield** [comment]
Duh, good point. Sorry.

So it's just the subsequent error that needs fixing.

**Comment by jkbonfield** [comment]
I found that I don't need a length for a contig in order to get bcftools to accept it and create BCFs.  Just the name alone is sufficient.

I wonder if we could create a tool that reads a VCF in its entirety and at the end spits out a new header, containing the minimal information needed to convert it to a BCF?  It won't be pretty, but in many cases we may not need to do this more than once or twice (eg 1000s of VCF.gz files all produced using the same pipeline).

I'd also urge you to put your weight behind making these fields mandatory in VCF v4.4!  You have a lot of experience in VCF and are officially a VCF maintainer, so 4.4 could benefit from your practical experience in developing and maintaining a VCF parser.  (We have a meeting this afternoon btw)

**Comment by pd3** [comment]
That's a neat idea of generating the header on the fly and adding it later. There is a problem with the annotation type: in order to proceed we declare it as String and write it out as such. Maybe a more useful tool would be to scan a VCF and create a header-like output with missing tags appended to the end. There could be a heuristics applied to prefill the type and the number of values, `bcftools reheader` could be then used afterwards to replace it.

**Comment by daviesrob** [comment]
Fixed by #1510
